libs/SmartSync/src/com/salesforce/androidsdk/smartsync/target/LayoutSyncDownTarget.java

```diff
@@ -0,0 +1,136 @@
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.salesforce.androidsdk.smartsync.target;
+
+import com.salesforce.androidsdk.rest.RestRequest;
+import com.salesforce.androidsdk.rest.RestResponse;
+import com.salesforce.androidsdk.smartsync.manager.SyncManager;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.IOException;
+import java.util.Set;
+
+/**
+ * Sync down target for object layouts. This uses the '/ui-api/layout' API to fetch object layouts.
+ *
+ * @author bhariharan
+ */
+public class LayoutSyncDownTarget extends SyncDownTarget {
+
+    public static final String SOBJECT_TYPE = "sobjectType";
+    public static final String LAYOUT_TYPE = "layoutType";
+
+    private String objectType;
+    private String layoutType;
+
+    /**
+     * Parameterized constructor.
+     *
+     * @param target JSON representation.
+     * @throws JSONException Exception thrown.
+     */
+    public LayoutSyncDownTarget(JSONObject target) throws JSONException {
+        super(target);
+        this.objectType = target.getString(SOBJECT_TYPE);
+        this.layoutType = target.getString(LAYOUT_TYPE);
+    }
+
+    /**
+     * Parameterized constructor.
+     *
+     * @param objectType Object type.
+     * @param layoutType Layout type.
+     */
+    public LayoutSyncDownTarget(String objectType, String layoutType) {
+        super();
+        this.queryType = QueryType.layout;
+        this.objectType = objectType;
+        this.layoutType = layoutType;
+    }
+
+    /**
+     * JSON representation of this target.
+     *
+     * @return JSON representation of this target.
+     * @throws JSONException Exception thrown.
+     */
+    public JSONObject asJSON() throws JSONException {
+        final JSONObject target = super.asJSON();
+        target.put(SOBJECT_TYPE, objectType);
+        target.put(LAYOUT_TYPE, layoutType);
+        return target;
+    }
+
+    @Override
+    public JSONArray startFetch(SyncManager syncManager, long maxTimeStamp) throws IOException, JSONException {
+        final RestRequest request = RestRequest.getRequestForObjectLayout(syncManager.apiVersion,
+                objectType, layoutType);
+        final RestResponse response = syncManager.sendSyncWithSmartSyncUserAgent(request);
+        final JSONArray records = new JSONArray();
+        records.put(response.asJSONObject());
```

Comment:
What's the shape of the response? What indexes should the soup they sync into have?
Should we provide some higher level (helper code) to setup that soup and the metadata syncs?
That could be the new MetaDataManager: the app would register interest for certain object type and we would create the soup and syncs and have methods to refresh the metadata. We could also have helper method to read it back from the store. Also, since the shape is known, we could have a typed object that is built from the json.

Reply:
For layouts, the shape can change based on the type of layout. I hesitate to strong-type this because more layouts could be added. The previous implementation had some strong typing since it was locked down on search layout.

Reply:
As for `describe`, it's really hard to tell which fields an app is interested in. The field list could be drastically different based on the use case of the app. If the app knows what fields they're looking for, it might be easier for the app to specify the soup index specs.

Reply:
There would be one soup for object infos and another for layouts.

For the object infos, there would be one row per object type which in the ui sdk object info response body is called apiName: so the object infos soup would only need a index on that field.

For layouts, should would the app retrieve the layout by id or should we add the object type in there and allow them to retrieve layout by object type+layout type+mode - in which case, those fields should be indexed.

Reply:
The shape of the layout info and object info is well defined.
See
* https://developer.salesforce.com/docs/atlas.en-us.uiapi.meta/uiapi/ui_api_responses_object_info.htm
* https://developer.salesforce.com/docs/atlas.en-us.uiapi.meta/uiapi/ui_api_responses_record_layout.htm

So we could base our classes on that.
Unlike data, metadata is quite static.

Reply:
Cool. I could add a thin helper layer on top of this that sets up the soup and provides utils to sync layout and metadata (the new `MetadataManager`). I'll do that in the next PR.